File: algorithms/python/Challenge_Largest_Divisible_Subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        if len(nums) < 2:
            return nums
        
        nums.sort()
        
        N = len(nums)
        c = [1 for _ in range(N)]
        dp = [0 for _ in range(N)]
        maxIdx = -1
        maxVal = -1
        for i in range(1, N):
            for j in range(0, i):
                if nums[i]%nums[j] == 0 and c[i] <= c[j] + 1:
                    c[i] = c[j] + 1
                    dp[i] = j
                    
            if c[i] > maxVal:
                maxIdx, maxVal = i, c[i]
                
        retList = []
        for _ in range(maxVal):
            retList.append(nums[maxIdx])
            maxIdx = dp[maxIdx]
        
        return retList

File: algorithms/java/Jun.2020.challenge/Challenge_Largest_Divisible_Subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        if(nums.length == 0)
            return new ArrayList<Integer>();
        else if(nums.length == 1)
            return new ArrayList<Integer>(List.of(nums[0]));
        
        Arrays.sort(nums);
        
        int N = nums.length;
        int[] c = new int[N];
        int[] dp = new int[N];
        Arrays.fill(c, 1);
        int maxVal = -1;
        int maxIdx = -1;
        
        for(int i = 1; i < N; i++) {
            for(int j = 0; j < i; j++) {
                if(nums[i]%nums[j] == 0 && c[i] <= c[j] + 1) {
                    c[i] = c[j]+1;
                    dp[i] = j;
                }
            }
            if(c[i] > maxVal) {
                maxVal = c[i];
                maxIdx = i;
            }
        }
        
        List<Integer> retList = new ArrayList<>();
        for(int i = 0; i < maxVal; i++) {
            retList.add(nums[maxIdx]);
            maxIdx = dp[maxIdx];
        }
        
        return retList;
    }
}
